HSDS/Reader/Reader.cpp

```diff
@@ -215,6 +215,128 @@ class Listener : public DDS::DataReaderListener
   typename OpenDDS::DCPS::DDSTraits<T>::DataReaderType::_var_type reader_;
 };
 
+template<typename T>
+class DdsListener : public DDS::DataReaderListener
+{
+public:
+  DdsListener(Application& application)
+    : application_(application)
+  {
+    DDS::DataReader_var reader = application.reader<T>();
+    reader_ = OpenDDS::DCPS::DDSTraits<T>::DataReaderType::_narrow(reader);
+    if (!reader_) {
+      ACE_ERROR((LM_ERROR, "ERROR: narrow failed!\n"));
+    }
+
+    reader_->set_listener(this, DDS::DATA_AVAILABLE_STATUS);
+    on_data_available(reader_);
+  }
+
+  void on_requested_deadline_missed(DDS::DataReader_ptr,
+                                    const DDS::RequestedDeadlineMissedStatus&)
+  {}
+
+  void on_requested_incompatible_qos(DDS::DataReader_ptr,
+                                     const DDS::RequestedIncompatibleQosStatus&)
+  {}
+
+  void on_sample_rejected(DDS::DataReader_ptr,
+                          const DDS::SampleRejectedStatus&)
+  {}
+
+  void on_liveliness_changed(DDS::DataReader_ptr,
+                             const DDS::LivelinessChangedStatus&)
+  {}
+
+  void on_data_available(DDS::DataReader_ptr)
+  {
+    ACE_GUARD(ACE_Thread_Mutex, g, application_.get_mutex());
```

Comment:
This can be moved to before `set_count()` is called at the end of this method to reduce the duration of holding `application_`'s mutex.

Reply:
I see. I'd prefer to have another mutex in this listener class to protect `keys_` to make it clear and separate from things protected by the application's mutex. But I get the point here.